### PR TITLE
Mm 47646 Global banner that indicate overage users

### PR DESCRIPTION
**Pull request by Afsoon**
#### Summary
Add a warning and error global banner that show the number of overage users.
Show only contact sales to admin who bought the license.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-47646

#### Related Pull Requests
- Has server changes https://github.com/mattermost/mattermost-server/pull/21638

#### Screenshots

Between 5% and 10% and admin who bought the license
image

Between 5% and 10% and admin who didn't buy the license
image

Over 10% and admin who bought the license
image

Over 10% and admin who didn't buy the license
image

#### Release Note
```release-note
Global banner for overage users in self hosted license
```

### CODE REVIEW

File: components/announcement_bar/overage_users_banner/index.tsx [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    const activeUsers = ((stats || {})[StatTypes.TOTAL_USERS]) as number || 0;
+    const {
+        isBetween5PercerntAnd10PercentPurchasedSeats,
+        isOver10PercerntPurchasedSeats,
+    } = calculateOverageUserActivated({
+        activeUsers,
+        seatsPurchased,
+    });
+    const prefixPreferences = isOver10PercerntPurchasedSeats ? 'error' : 'warn';
+    const preferenceName = `${prefixPreferences}_overage_seats_${seatsPurchased}`;
```
Comment (neallred) [inline components/announcement_bar/overage_users_banner/index.tsx:55]: Neat. If I understand, this will make it so that each time they add a new user, they will re-see the banner until they dismiss it again (and if they already dismissed it and go down to a number they have already seen, they won't see it again).

File: components/announcement_bar/overage_users_banner/index.tsx [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+    const overageByUsers = activeUsers - seatsPurchased;
+
+    const isOverageState = isBetween5PercerntAnd10PercentPurchasedSeats || isOver10PercerntPurchasedSeats;
+
+    if (!isAdmin || !isOverageState || isCloud || adminHasWatchIt({isWarningBanner: isBetween5PercerntAnd10PercentPurchasedSeats, overagePreferences, preferenceName})) {
+        return null;
+    }
+
+    const handleClose = () => {
```
Comment (neallred) [inline components/announcement_bar/overage_users_banner/index.tsx:65]: Optional: The combo of overagePreferenves, the find helper, the and `dispatch(savePreferences` all works well. If you'd like, some of it *might* be simpler if you used the [`usePreference`](https://github.com/mattermost/mattermost-webapp/blob/master/components/common/hooks/usePreference.ts) hook.

**Review by neallred — APPROVED** [review]
LGTM, thanks @Afsoon !

File: e2e/cypress/tests/integration/enterprise/overage_users/info_overage_users_spec.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        cy.apiInitSetup().then(({user, offTopicUrl}) => {
+            url = offTopicUrl;
+            createdUser = user;
+            cy.apiAdminLogin();
+            cy.apiDeleteLicense();
+            cy.visit(url);
+        });
+    });
+
+    it('should show the banner with CTA because it\'s the admin who purchased the license', () => {
```
Comment (yasserfaraazkhan) [inline e2e/cypress/tests/integration/enterprise/overage_users/info_overage_users_spec.ts:59]: thank you for the tests @Afsoon .
- Do you think we can merge both files ? (It will be less load in splitting specs CI runners).
- in case we are using this `cy.apiDeleteLicense()`, can we keep it outside /cloud folder. It kinda blocks the other cloud tests that require license. I can help if you need anything.

**Review by yasserfaraazkhan — CHANGES_REQUESTED** [review]

File: components/announcement_bar/overage_users_banner/index.tsx (hunk 1 above)
Comment (Afsoon) [reply components/announcement_bar/overage_users_banner/index.tsx:55]: Exactly, and in the second scenario only when they are over the 10% threshold then they will see it again.

**Comment by neallred** [comment]
As a heads up/FYI: To enable testing this we need a license with few users. To use a test license, we need the server version to have a publicKey matching the test license generator's private key. To do that, we need an accompanying mattermost-server change, which I've done here: https://github.com/mattermost/mattermost-server/tree/mm-47646-dynamic-banner-system-console-license-screen. Doing an `/update-branch` in the hopes of getting the matching server version pulled in. If it doesn't, we might need to delete the current spinwick and spin up a new one.

**Comment by neallred** [comment]
Didn't take. Will try deleting and updating.

**Comment by neallred** [comment]
@yasserfaraazkhan @stevemudie For whatever reason, the spinwick we create from mattermost-webapp isn't picking up the server changes, but we can use the dummy server side spinwick I spun up because it has the right license code as well as these front end changes. Details:

URL: https://mattermost-server-pr-21638.test.mattermost.cloud


Account Type | Username | Password
-- | -- | --
Admin | sysadmin | Sys@dmin123
User | user-1 | User-1@123

**Comment by stevemudie** [comment]
Thanks for your help on this @neallred!

@yasserfaraazkhan I've played with it a bit but will let you take it from here.

You can activate and deactivate users in the System Console as needed to see the banner respond but you need to refresh to get it to update.

I did notice that just 1 user over triggers the red banner which isn't as it should be I don't think.

File: components/announcement_bar/overage_users_banner/index.tsx (hunk 1 above)
Comment (marianunez) [reply components/announcement_bar/overage_users_banner/index.tsx:55]: Each time they add a new user it will show up again? The preference seems to always be the same for the combination `warn` + `seatsPurchased`

File: components/announcement_bar/overage_users_banner/index.tsx [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+import './overage_users_banner.scss';
+
+type AdminHasWatchItArgs = {
+    preferenceName: string;
+    overagePreferences: PreferenceType[];
+    isWarningBanner: boolean;
+}
+
+const adminHasWatchIt = ({preferenceName, overagePreferences, isWarningBanner}: AdminHasWatchItArgs): boolean => {
```
Comment (marianunez) [inline components/announcement_bar/overage_users_banner/index.tsx:26]: nit: given that the admin has to explicitly dismiss it, should we rename this so that it's not confused that it's only when they see it once :

```suggestion
const adminHasDismissed = ({preferenceName, overagePreferences, isWarningBanner}: AdminHasWatchItArgs): boolean => {
```

File: components/announcement_bar/overage_users_banner/index.tsx [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+            showCloseButton={isBetween5PercerntAnd10PercentPurchasedSeats}
+            onButtonClick={handleClick}
+            modalButtonText={cta}
+            modalButtonDefaultText={cta}
+            message={message}
+            showLinkAsButton={true}
+            isTallBanner={true}
+            icon={<i className='icon icon-alert-outline'/>}
+            handleClose={handleClose}
+            showCTA={license.Email === user.email}
```
Comment (marianunez) [inline components/announcement_bar/overage_users_banner/index.tsx:105]: Why are we limiting the purchasing admin to Contact Sales? May times the emails used in the license are things like `admin@mattermost.com` and don't match a real admin user

**Review by marianunez — APPROVED** [review]
Great work @Afsoon! Minor comments below, mostly wondering why we limit the Contact sales to the purchasing admin

File: components/announcement_bar/overage_users_banner/index.tsx [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+// See LICENSE.txt for license information.
+import React from 'react';
+import {FormattedMessage, useIntl} from 'react-intl';
+import {useDispatch, useSelector} from 'react-redux';
+
+import {getCurrentUser, isCurrentUserSystemAdmin} from 'mattermost-redux/selectors/entities/users';
+import {GlobalState} from 'types/store';
+import {getLicense} from 'mattermost-redux/selectors/entities/general';
+import AnnouncementBar from 'components/announcement_bar/default_announcement_bar';
+import {calculateOverageUserActivated} from 'utils/overage_team';
```
Comment (marianunez) [inline components/announcement_bar/overage_users_banner/index.tsx:11]: Is this a new file? Doesn't seem to be part of the PR?

**Comment by Afsoon** [comment]
> Thanks for your help on this @neallred!
> 
> @yasserfaraazkhan I've played with it a bit but will let you take it from here.
> 
> You can activate and deactivate users in the System Console as needed to see the banner respond but you need to refresh to get it to update.
> 
> I did notice that just 1 user over triggers the red banner which isn't as it should be I don't think.

We have to change the license for 20 users, because 10 have an overlap between 5% and 10%. I'm rounding up the decimals. I will generate a new one with 20 seats. cc @yasserfaraazkhan

File: components/announcement_bar/overage_users_banner/index.tsx (hunk 6 above)
Comment (Afsoon) [reply components/announcement_bar/overage_users_banner/index.tsx:11]: Yes, but this PR is created from another branch. https://github.com/mattermost/mattermost-webapp/pull/11482/files#diff-d01376ffc6b70925fed81daa8f355304ae72969f3923a6227017572c1b24f6f2R2185

File: components/announcement_bar/overage_users_banner/index.tsx (hunk 1 above)
Comment (Afsoon) [reply components/announcement_bar/overage_users_banner/index.tsx:55]: it won't show up again, the warning state. Only when they buy new seats and reach overage users again.

File: components/announcement_bar/overage_users_banner/index.tsx (hunk 5 above)
Comment (Afsoon) [reply components/announcement_bar/overage_users_banner/index.tsx:105]: If we're going to have a lot of false negatives, I will remove it then.

File: components/announcement_bar/overage_users_banner/index.tsx (hunk 1 above)
Comment (marianunez) [reply components/announcement_bar/overage_users_banner/index.tsx:55]: Should we use the license id here instead to be more precise? There is a chance they purchase a new license with the same amount of seats

File: components/announcement_bar/overage_users_banner/index.tsx (hunk 5 above)
Comment (marianunez) [reply components/announcement_bar/overage_users_banner/index.tsx:105]: Well I was just wondering if this was part of the requirements in the design or where this came from?

Comment (Afsoon) [reply components/announcement_bar/overage_users_banner/index.tsx:105]: This check came from the new design for the global banner; reading the document attached to the epic, I can't see any requirements.

**Comment by yasserfaraazkhan** [comment]
Thank you @Afsoon ,
The changes are working as expected.

Just wanted to check, is there any precedence on which banner should be displayed ?
Currently, below banner supersedes the true up banner.

Screenshot 2022-11-14 at 10 45 59 PM

**Review by yasserfaraazkhan — APPROVED** [review]
https://github.com/mattermost/mattermost-test-management/pull/10
I'll push the above test to jira.

I've tested the changes on PR and its working fine. Just had curious question as stated above.
